### ISSUE TITLE
Allow ignoring all attributes

### ISSUE THREAD

**Issue by ismay** [edvardchen/eslint-plugin-i18next#39]
It would be nice to have an explicit way to ignore all attributes as all attribute warnings have been false positives for me so far. Currently I use: 

```
        'i18next/no-literal-string': [
            'error',
            { markupOnly: true, onlyAttribute: [""] },
        ],
```

But that's a rather confusing way of configuring this in my opinion.

**Comment by edvardchen** [comment]
@ismay we could add another option like `ignoreAllAttributes` but then your config would look like:

```js
        'i18next/no-literal-string': [
            'error',
            { markupOnly: true, ignoreAllAttributes: true },
        ],
```
Still awful 

Like I said in [this issue](https://github.com/edvardchen/eslint-plugin-i18next/issues/17#issuecomment-636488165), I hate to add two options.

So, I proposal to extend the `markupOnly` option:

1. `markupOnly: 1` that just validate the text nodes only, ignore all markups' attributes
2. `markupOnly: 2` that validate the text nodes and all markups' attributes

How do you think?

**Comment by ismay** [comment]
Yeah, tricky. I understand your point. I think that the:

```
markupOnly: 1
markupOnly: 2
```

would be a bit confusing though. I think for me, something in this direction would make sense:

```js
// This will lint both markup and attributes
markup: true,
// could also be written as
markup: [true, { attributes: true }],

// This will lint markup but ignore attributes
markup: [true, { attributes: false }],

// This will lint markup and only the foo and bar attributes
markup: [true, { attributes: ['foo', 'bar'] }],

// Turn on only markup
markup: [true, { onlyMarkup: true }],

// Ignore specific attributes
markup: [true, { ignoreAttributes: ['foo', 'bar'] }],
```

Though that's a breaking change, so you might want to give that some time to see whether it covers all usecases (and I'm not a 100% satisfied with the above). But I think with the current configuration I find it a bit tricky to implement this change in a way that's not confusing.

**Comment by edvardchen** [comment]
You make me review current options. We used to add options constantly for solving issues. We should do better.

For example, `onlyAttribute` and `ignoreAttribute` are a pairing and look structureless and verbose.

So I would like to re-design the interface, like:

```json
{
  // default is textOnly, can be textOnly, markupOnly or global
  "mode": "textOnly",
  "words": { "include": [], "exclude": [] },
  "functions": { "include": [], "exclude": [] },
  "templates": true, // validate string template or not
  "markups": {
    "attribute": { "exclude": [], "include": [] }
  },
  "properties": { "include": [], "exclude": [] },
  "components": { "include": [], "exclude": [] }
}
```

### `exclude` and `include`

The `exclude` and `include` were borrowed from [webpack ecosystem](https://webpack.js.org/configuration/). Instead of applying them to filter files, we use them to filter text.

We can do:

1. match all -- just omit two options
2. ignore all -- `"exclude": [".+"]`
3. match some and ignore some -- `"exclude": ["foo"],"include": ["bar.+"]`, both conditions must be satisfied

### mode

Validate only text nodes in markups:

```yaml
rules:
  i18next/no-literal-string:
    - 1
    - mode: textOnly
      # words:
      #   exclude:
      #   include:
```

Validate text and attributes in markups:

```yaml
rules:
  i18next/no-literal-string:
    - 1
    - mode: markupOnly
      # words:
      #   exclude:
      #   include:
      # markups:
      #   attribute:
      #     exclude:
      #     include:
```

Validate all literal strings in whole project:

```yaml
rules:
  i18next/no-literal-string:
    - 1
    - mode: global
      # all other options
```

@ismay tell me your thought

**Comment by ismay** [comment]
Interesting, yes I personally do think that that is a step in the right direction. I would talk to other people as well before making this decision though, just to make sure that everyone will be happy with the changes. And maybe then release it as an alpha first, to test the waters.

Couple more detailed points:

- Instead of calling it `jsx text` (or `textOnly`), maybe `children` / `childrenOnly` would be more clear.
- Instead of the `words` option, would `text` maybe be more appropriate? Since it's not always words that you're including or excluding, but it's always going to be text.
- Instead of `templates`, maybe `validateTemplates` would be more clear. Since you're using nouns as the include / exclude options. Plus `validateTemplates` tells me what it does more clearly
- `markups` I would change to `markup`. There's no plural as far as I'm aware.

This is also just my personal opinion, so do check if this fits with other people's usage of this plugin. Hope this helps!

**Comment by edvardchen** [comment]
`childrenOnly` may be more specified but not instinctive enough. I prefer `textNodesOnly`

```jsonc
{
  // default is textNodesOnly, can be markupOnly or global
  "mode": "global",
  // same for react and vue
  "markup": {
    "attributes": { "exclude": [], "include": [] },
    "elements": { "include": [], "exclude": [] }
  },
  "text": { "include": [], "exclude": [] },
  "functions": { "include": [], "exclude": [] },
  "validateTemplates": true, // validate string templates or not
  "object": {
    "properties": { "include": [], "exclude": [] }
  }
}
```

**Comment by ismay** [comment]
> `childrenOnly` may be more specified but not instinctive enough. I prefer `textNodesOnly`
> 
> ```js
> {
>   // default is textNodesOnly, can be markupOnly or global
>   "mode": "global",
>   // same for react and vue
>   "markup": {
>     "attributes": { "exclude": [], "include": [] },
>     "elements": { "include": [], "exclude": [] }
>   },
>   "texts": { "include": [], "exclude": [] },
>   "functions": { "include": [], "exclude": [] },
>   "validateTemplates": true, // validate string templates or not
>   "object": {
>     "properties": { "include": [], "exclude": [] }
>   }
> }
> ```

Apologies for the late reply. That looks good! Two minor remarks:

- I'd use `text` as key instead of `texts`. Same as my earlier comment on `markup`, I don't think the plural is appropriate in this context.
- Opposite comment for `object`, you could use the plural there: `objects`

```
{
  "mode": "global",
  "markup": {
    "attributes": { "exclude": [], "include": [] },
    "elements": { "include": [], "exclude": [] }
  },
  "text": { "include": [], "exclude": [] },
  "functions": { "include": [], "exclude": [] },
  "validateTemplates": true,
  "objects": {
    "properties": { "include": [], "exclude": [] }
  }
}
```

Let me know what you think!

**Comment by edvardchen** [comment]
Thank you and forgive my poor English

After days of reflection, I have some new thoughts.

Maybe this plugin made a wrong assumption that people would like to find out all hand-written strings. Many people just care more about the hand-written strings in the markup (JSX or Vue template) than those elsewhere. And they may think the ‘pollution’ through variables like (`<div>{msg}/</div>`) is acceptable.

I think we should split the current rule `no-literal-string` into two:

1. `no-text-in-markup` validate literal strings in markup, just the text nodes and attributes with literal-string-value. The configuration is simple:

   ```jsonc
   {
     "attributes": { "exclude": [], "include": [] },
     "elements": { "include": [], "exclude": [] }
   }
   ```

2. `no-magic-strings` validate all strings that users initialize. It works like [no-magic-numbers](https://eslint.org/docs/rules/no-magic-numbers). This rule provides many options:

   ```jsonc
   {
     "text": { "include": [], "exclude": [] },
     "functions": { "include": [], "exclude": [] },
     "validateTemplates": true,
     "objects": {
       "properties": { "include": [], "exclude": [] }
     }
   }
   ```
@ismay expect your reply

**Comment by ismay** [comment]
> Thank you and forgive my poor English

No worries, your English is great! It's not my native language either, so feel free to correct me if I'm wrong with some of my suggestions.

> After days of reflection, I have some new thoughts.
> 
> Maybe this plugin made a wrong assumption that people would like to find out all hand-written strings. Many people just care more about the hand-written strings in the markup (JSX or Vue template) than those elsewhere. And they may think the ‘pollution’ through variables like (`<div>{msg}/</div>`) is acceptable.

Interesting. Yes I think that that's a good approach. Looking at the level of complexity for the configuration of the existing rule, it seems likely that it's doing too much. Splitting it up seems like a good way of simplifying it. I was thinking about this, and I think I'd suggest a couple alterations:

> I think we should split the current rule `no-literal-string` into two:
> 
>     1. `no-text-in-markup` validate literal strings in markup, just the text nodes and attributes with literal-string-value. The configuration is simple:
>        ```js
>        {
>          "attributes": { "exclude": [], "include": [] },
>          "elements": { "include": [], "exclude": [] }
>        }
>        ```

So I was thinking, the goal of this rule in the end, is to catch untranslated string in markup. I'm wondering if maybe `no-untranslated-strings-in-markup` would convey the intention of the rule better.

Also, allowing for the `ignoreCallee` option of the current plugin seems nice here, so that you can supress the warning if the string is correctly translated.

>     2. `no-magic-strings` validate all strings that users initialize. It works like [no-magic-numbers](https://eslint.org/docs/rules/no-magic-numbers). This rule provides many options:
>        ```js
>        {
>          "text": { "include": [], "exclude": [] },
>          "functions": { "include": [], "exclude": [] },
>          "validateTemplates": true,
>          "objects": {
>            "properties": { "include": [], "exclude": [] }
>          }
>        }
>        ```

I think I understand the intention here, but I think that the definition for magic string is slightly different than the definition for magic number: https://en.wikipedia.org/wiki/Magic_string. I was thinking about this and first I thought `no-string-literals` could work, but then as a user I would expect that to lint markup as well, so that doesn't seem appropriate.

So I was thinking, maybe you could offer the same level of granularity as the `no-untranslated-strings-in-markup`, and go with `no-untranslated-strings-in-text`, `no-untranslated-strings-in-functions`, etc. That would allow you to keep all of the rules very simple.

What do you think?

**Comment by edvardchen** [comment]
> So I was thinking, the goal of this rule in the end, is to catch untranslated string in markup. I'm wondering if maybe `no-untranslated-strings-in-markup` would convey the intention of the rule better.

The name `no-untranslated-strings-in-markup` looks better.

> Also, allowing for the `ignoreCallee` option of the current plugin seems nice here, so that you can supress the warning if the string is correctly translated.
>
> I think I understand the intention here, but I think that the definition for magic string is slightly different than the definition for magic number: https://en.wikipedia.org/wiki/Magic_string. I was thinking about this and first I thought `no-string-literals` could work, but then as a user I would expect that to lint markup as well, so that doesn't seem appropriate.

Sorry. I misunderstand `magic-string`. In eslint's world, rules don't interact. They don't know each other.

If we add another rule `no-untranslated-strings-in-markup` and we want `no-string-literals` to link markup as well, then `no-string-literals` should contains all logic of `no-untranslated-strings-in-markup`. And there is no easy way to extract common logic from them.

> So I was thinking, maybe you could offer the same level of granularity as the `no-untranslated-strings-in-markup`, and go with `no-untranslated-strings-in-text`, `no-untranslated-strings-in-functions`, etc. That would allow you to keep all of the rules very simple.

We can't do that for the same reason as above. Thinking about this example:

```javascript
// code
const foo = {
  bar() {
    alert('hello');
  },
};

// rule config
{
  "no-untranslated-strings-in-callee": 2,
  "no-untranslated-strings-in-object-properties": [1, {
    "exclude": ["bar"]
  }]
}
```

We should not yell about the `alert('hello');` but we would if we divide current plugin at the level of granularity above.

Seems that we back to the start. Cannot split the original rule `no-literal-string`

**Comment by edvardchen** [comment]
@ismay   try the prerelease version v6.0.0-0. It will only validate the plain text in JSX markup.

**Comment by ismay** [comment]
Will do, thanks!
